### PR TITLE
Notify post author about new comments by sending email

### PR DESCRIPTION
**Pull request by voronkovich**
See https://github.com/symfony/symfony-demo/pull/420

### CODE REVIEW

**Comment by javiereguiluz** [comment]
I like this feature proposal ... but what about replacing the private method in the controller by an event listener + custom event trigger? That would allow us to show custom events in action.

**Comment by voronkovich** [comment]
@javiereguiluz, I've added an event listener and custom event trigger. BTW, I found that we have an example of using mailer in the `ListUsersCommand` https://github.com/symfony/symfony-demo/blob/master/src/AppBundle/Command/ListUsersCommand.php#L131

File: app/config/services.yml [hunk 1]
```diff
@@ -31,9 +31,15 @@ services:
...
             - { name: kernel.event_listener, event: kernel.request, method: onKernelRequest }
 
-    # Event subscribers are similar to event listeners but they don't need to add
-    # a separate tag for each listened event. Instead, the PHP class of the event
-    # subscriber includes a method that returns the list of listened events.
+    app.comment_notification:
+        class: AppBundle\EventListener\CommentNotificationListener
+        arguments: ['@mailer', '@router', '@translator', '%app.notifications.email_sender%']
+        tags:
+            - { name: kernel.event_listener, event: comment.created, method: onCommentCreated }
```
Comment (javiereguiluz) [inline app/config/services.yml]: I have two questions about the `onCommentCreated` method name:

1) When the method name is `on + camelCasedEventName`, the `method` parameter is optional. Should we add it anyway?
2) Some people like method names that describe the listened event (e.g. `on + EventName`) and others prefer semantic method names that describe their purposes (e.g. `notifyCommentAdded()`). Which is your preference?

File: src/AppBundle/EventListener/CommentNotificationListener.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @var string
+     */
+    private $sender = 'symfony-demo@localhost';
```
Comment (bocharsky-bw) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: What about to make it as a parameter in `config.yml`?

Comment (javiereguiluz) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: Yes. It's already passed to the constructor of the service, so I'd say: let's remove this private `$sender` property and just use the container parameter.

Comment (voronkovich) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: @javiereguiluz, @bocharsky-bw, do you suggest to inject container into the listener to get the `sender` parameter like there https://github.com/symfony/symfony-demo/blob/master/src/AppBundle/Command/ListUsersCommand.php#L138
As you can see in the listener definition those parameter (`app.notifications.email_sender`) is already injected.

Comment (stof) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: no, keep the property here, but removes its default value and make it a mandatory constructor argument

File: app/config/services.yml [hunk 3]
```diff
@@ -31,9 +31,15 @@ services:
         tags:
             - { name: kernel.event_listener, event: kernel.request, method: onKernelRequest }
 
-    # Event subscribers are similar to event listeners but they don't need to add
-    # a separate tag for each listened event. Instead, the PHP class of the event
-    # subscriber includes a method that returns the list of listened events.
+    app.comment_notification:
+        class: AppBundle\EventListener\CommentNotificationListener
+        arguments: ['@mailer', '@router', '@translator', '%app.notifications.email_sender%']
```
Comment (stof) [inline app/config/services.yml]: where is the ``app.notifications.email_sender`` parameter initialized ?

Comment (stof) [inline app/config/services.yml]: OK, it is already part of the codebase although not used for notifications today

File: src/AppBundle/EventListener/CommentNotificationListener.php (hunk 2 above)
Comment (voronkovich) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: Ok. I've removed the default sender's value.

File: app/config/services.yml [hunk 4]
```diff
@@ -31,9 +31,18 @@ services:
...
+        # There is an optional tag attribute called "method" which defines which method to execute when the event is triggered.
+        # By default the name of the method is on + "camel-cased event name".
+        # If the event is "comment.created" the method executed by default is onCommentCreated().
```
Comment (voronkovich) [inline app/config/services.yml]: @javiereguiluz,  I've added a note about listener method's default name.

File: src/AppBundle/EventListener/CommentNotificationListener.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        $comment = $event->getSubject();
+        $post = $comment->getPost();
+
+        $linkToPost = $this->urlGenerator->generate('blog_post', ['slug' => $post->getSlug()], UrlGeneratorInterface::ABSOLUTE_URL);
+
+        $subject = $this->translator->trans('post.recieved_comment');
+        $body = $this->translator->trans('post.recieved_comment_message', [
+            '%title%' => $post->getTitle(),
+            '%link%' => $linkToPost.'#comment_'.$comment->getId(),
```
Comment (javiereguiluz) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: @voronkovich first, thank you for your patience with this code review 😄 

I have a last comment about your PR. What if we use the shinny new [URL fragment support](http://symfony.com/blog/new-in-symfony-3-2-routing-improvements#added-support-for-url-fragments) in Symfony 3.2 to forget about concatenating strings to define fragments?

Comment (voronkovich) [inline src/AppBundle/EventListener/CommentNotificationListener.php]: @javiereguiluz, it's a good idea! I've updated an url generating to use the `_fragment` parameter.

File: app/Resources/translations/messages.en.xlf [hunk 6]
```diff
@@ -276,6 +276,14 @@
                 <source>post.deleted_successfully</source>
                 <target>Post deleted successfully!</target>
             </trans-unit>
+            <trans-unit id="post.comment_added">
```
Comment (yceruto) [inline app/Resources/translations/messages.en.xlf]: Just for consistency matter (with event name and method name) shouldn't be named `post.comment_created` instead?

Comment (voronkovich) [inline app/Resources/translations/messages.en.xlf]: @yceruto, I think you're right! I've changed messages names and added a more appropriate prefix for them.

**Comment by javiereguiluz** [comment]
@voronkovich another great feature added to the Demo app! Thank you very much for contributing this.

**Pull request merged**